### PR TITLE
Restrict `Allocator` impl to `&'static A`

### PR DESCRIPTION
**Pull request by djkoloski**
Arbitrary `&'_ Allocator`s cannot satisfy the safety requirement that
memory blocks returned by an `Allocator` must remain valid until the
instance and all of its clones are dropped. `&'static Allocator`s can
still satisfy this safety requirement because their lifetime ensures
that they are never moved or dropped.

See #90822 and [this comment](https://github.com/rust-lang/rust/pull/90822#issuecomment-1041112411) in particular for additional context.

### CODE REVIEW

**Comment by apiraino** [comment]
I think this `t-libs` rather than `t-compiler`

@rustbot label -t-compiler +t-libs

**Comment by yaahc** [comment]
I'm reassigning this PR because I'm taking a break from the review rotation for a little while. Thank you for your patience.

r? rust-lang/libs-api

**Comment by joboet** [comment]
Couldn't the documentation just be updated to clarify that the allocated blocks are only valid for `'a where A: 'a` (not sure how to write that in plain English)? That would match the current assumptions in `Box::leak` and still allow allocating multiple structures with the same allocator, which therefore can be stack-based.

**Comment by djkoloski** [comment]
#94114 has a more extensive change that pursues a similar goal.

**Comment by JohnCSimon** [comment]
> https://github.com/rust-lang/rust/pull/94114 has a more extensive change that pursues a similar goal.
Triage: 
@djkoloski what should happen to this PR? Maybe close it?

**Comment by djkoloski** [comment]
This PR is meant as a minimal fix to unsoundness in the changed impl. The other PR is a more extensive change that provides more flexibility, so if we'd like to go forward with #94114 then I can close this one.

**Comment by pitaj** [comment]
Ping from triage. @djkoloski looks like your other PR has more traction, do you want to close this one?

**Comment by djkoloski** [comment]
Note that we did not go forward with the other PR. We should probably focus on merging this and closing the soundness hole.

**Comment by joboet** [comment]
I'm nominating this for team discussion. The question is whether to maintain the current guarantee that
> Memory blocks returned from an allocator that are currently allocated [...] retain their validity while [...] at least one of the instance and all of its clones has not been dropped

which would require allocators to have `'static` lifetime, or whether to relax this guarantee to apply only to `'static` allocators (this is what is currently done in practice with `Box::leak`. 

@rustbot label +I-libs-api-nominated

**Comment by Amanieu** [comment]
Closing in favor of #118890.
